File: fixtures/projects/e2e-python/src/middleware/cors.py
from dataclasses import dataclass

from src.constants.http import HTTP_NO_CONTENT


@dataclass(slots=True)
class CorsPolicy:
    allow_origin: str
    allow_methods: tuple[str, ...]


DEFAULT_POLICY = CorsPolicy(allow_origin="*", allow_methods=("GET", "POST", "OPTIONS"))


def build_preflight_headers(
    origin: str, policy: CorsPolicy = DEFAULT_POLICY
) -> tuple[int, dict[str, str]]:
    headers = {
        "access-control-allow-origin": origin or policy.allow_origin,
        "access-control-allow-methods": ", ".join(policy.allow_methods),
    }
    return HTTP_NO_CONTENT, headers

File: fixtures/projects/e2e-app/src/middleware/cors.ts
import { DEFAULT_PAGE_SIZE } from "../constants/app";

export type CorsOptions = {
	origins: string[];
	methods: string[];
	allowCredentials: boolean;
};

export function createCorsHeaders(origin: string | undefined, options: CorsOptions): Record<string, string> {
	const allowedOrigin = origin && options.origins.includes(origin) ? origin : options.origins[0] ?? "*";
	return {
		"access-control-allow-origin": allowedOrigin,
		"access-control-allow-methods": options.methods.join(", "),
		"access-control-allow-credentials": String(options.allowCredentials),
		"x-default-page-size": String(DEFAULT_PAGE_SIZE),
	};
}

export function shouldHandlePreflight(method: string, headers: Record<string, string | undefined>): boolean {
	return method.toUpperCase() === "OPTIONS" && Boolean(headers["access-control-request-method"]);
}